overlord/snapstate/handlers.go

```diff
@@ -1934,6 +1934,74 @@ func (m *SnapManager) doUnlinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	return err
 }
 
+func (m *SnapManager) undoUnlinkSnap(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	snapsup, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+
+	// XXX: is it safe (can panic)?
+	isInstalled := snapst.IsInstalled()
+	if !isInstalled {
+		return fmt.Errorf("internal error: snap %q not installed", snapsup.InstanceName())
+	}
+
+	info, err := snapst.CurrentInfo()
+	if err != nil {
+		return err
+	}
+
+	deviceCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+
+	// undo here may be part of failed snap remove change, in which case a later
+	// "clear-snap" task could have been executed and some or all of the
+	// data of this snap could be lost. If that's the case, then we should not
+	// enable the snap back.
+	// XXX: should make an exception for snapd/core?
+	place := snapsup.placeInfo()
+	for _, dir := range []string{place.DataDir(), place.CommonDataDir()} {
+		if exists, _, _ := osutil.DirExists(dir); !exists {
```

Comment:
it this situation we need to mark the snap broken somehow ?  wonder if we should do it here, or where?

Reply:
Yes, I was looking at broken flag but we only have it on the snap.Info, set in readInfo, that's too unrelated (and it's about being broken for other reasons). If we really need to mark it broken, then we should discuss how, a new flag on the state would affect lots of places I suppose.

Reply:
yes, I was misremembering,  I thought we had a Broken concept at the SnapState level but we don't, let's just leave a TODO here.

Reply:
Done